include/xtensor/xsort.hpp

```diff
@@ -144,6 +144,148 @@ namespace xt
+                                          std::size_t(1), std::multiplies<>());
+                data_secondary_stride = static_cast<std::ptrdiff_t>(data.strides()[data.dimension() - 2]);
+                inds_secondary_stride = static_cast<std::ptrdiff_t>(inds.strides()[inds.dimension() - 2]);
+            }
+            else
+            {
+                n_iters = std::accumulate(data.shape().begin() + 1, data.shape().end(),
+                                          std::size_t(1), std::multiplies<>());
+                data_secondary_stride = static_cast<std::ptrdiff_t>(data.strides()[1]);
+                inds_secondary_stride = static_cast<std::ptrdiff_t>(inds.strides()[1]);
+            }
+
+            std::ptrdiff_t data_offset = 0;
+            std::ptrdiff_t inds_offset = 0;
+
+            for (std::size_t i = 0; i < n_iters; ++i, data_offset += data_secondary_stride,
+                    inds_offset += inds_secondary_stride)
+            {
+                auto comp = [&data, &data_offset](std::size_t x, std::size_t y) {
+                    return (*(data.data() + data_offset + x) <
+                            *(data.data() + data_offset + y));
+                };
+                std::iota(inds.data() + inds_offset, inds.data() + inds_offset + inds_secondary_stride, 0);
+                std::sort(inds.data() + inds_offset, inds.data() + inds_offset + inds_secondary_stride, comp);
+            }
+        }
+    }
+
+    template <class E>
+    auto argsort(const xexpression<E>& e, placeholders::xtuph /*t*/)
+    {
+        using value_type = typename E::value_type;
+        using result_type = typename detail::argsort_result_type<E>::type;
+
+        const auto de = e.derived_cast();
+
+        result_type result;
+        result.resize({de.size()});
+        auto comp = [&de](std::size_t x, std::size_t y) {
+            return de[x] < de[y];
+        };
+        std::iota(result.begin(), result.end(), 0);
+        std::sort(result.begin(), result.end(), comp);
+
+        return result;
+    }
+
+    /**
+     * Argsort xexpression (optionally along axis)
+     * Performs an indirect sort along the given axis. Returns an xarray
+     * of indices of the same shape as e that index data along the given axis in
+     * sorted order.
+     *
+     * @param e xexpression to argsort
+     * @param axis axis along which argsort is performed
+     *
+     * @return argsorted index array
+     */
+    template <class E>
+    auto argsort(const xexpression<E>& e, std::size_t axis)
+    {
+        using eval_type = typename E::temporary_type;
+        using value_type = typename E::value_type;
+        using result_type = typename detail::argsort_result_type<E>::type;
+
+        const auto& de = e.derived_cast();
+
+        if (de.dimension() == 1)
+        {
+            return argsort(de, xnone());
+        }
+
+        eval_type ev;
+
+        if (axis != detail::leading_axis(ev))
```

Comment:
`ev.layout()` might return a value different from `de.layout()` if the compile-time layout of `e` is `dynamic`. So the following might be better:

```cpp
if(axis != detail::leading_axis(de))
```

Besides, this gives you the opportunity to declare `ev` later, avoiding a temporary when assigning to it.